LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -30,96 +39,110 @@ internal sealed class ConcentratorDeduplication<T> :
 
         public ConcentratorDeduplication(
             IMemoryCache cache,
-            WebSocketWriterRegistry<StationEui, string> socketRegistry,
-            ILogger<IConcentratorDeduplication<T>> logger)
+            ILogger<IConcentratorDeduplication> logger)
         {
             this.cache = cache;
-            this.socketRegistry = socketRegistry;
             this.logger = logger;
         }
 
-        public bool ShouldDrop(T frame, StationEui stationEui)
+        public async Task<Result> CheckDuplicateAsync(LoRaRequest loRaRequest, LoRaDevice? loRaDevice)
         {
-            if (frame == null) throw new ArgumentNullException(nameof(frame));
+            _ = loRaRequest ?? throw new ArgumentNullException(nameof(loRaRequest));
+            if (loRaDevice is null && loRaRequest.Payload is LoRaPayloadData)
+                throw new ArgumentNullException(nameof(loRaDevice));
 
-            var key = CreateCacheKey(frame);
+            var key = CreateCacheKey(loRaRequest);
+            var stationEui = loRaRequest.StationEui;
 
             StationEui previousStation;
-            lock (this.cache)
+            await this.cacheSemaphore.WaitAsync();
+            try
             {
                 if (!this.cache.TryGetValue(key, out previousStation))
                 {
-                    AddToCache(key, stationEui);
-                    return false;
+                    _ = this.cache.Set(key, stationEui, new MemoryCacheEntryOptions()
+                    {
+                        SlidingExpiration = DefaultExpiration
+                    });
+                    return Result.NotDuplicate;
                 }
             }
+            finally
+            {
+                _ = this.cacheSemaphore.Release();
+            }
 
-            if (previousStation == stationEui)
+            if (loRaRequest.Payload.RequiresConfirmation() && previousStation == stationEui)
```

Comment:
nit: I would turn around the 2 conditions. 
1. it makes it clearer that this condition only applies to the same station that is also stored in the cache
2. the validation of the stationEUI is quicker and all subsequent messages, don't need to validate the fact if it RequiresConfirmation.

Reply:
We removed the requires confirmation check since we want to deduplicate also unconfirmed messages.